### PR TITLE
Introduce the quarkus.container-image.image property

### LINKED ISSUE #8588: Support GitLab CI_REGISTRY_IMAGE variable for container build.
**Description**

In GitLab, projects have concrete image names that matches projects. The image name is provided to build script in environment variable CI_REGISTRY_IMAGE. However, variable also contains repository name so it does not work with container-image (I tested with JIB):

CI_REGISTRY_IMAGE=gitlab.company.hr:1111/group/subgroup/project

If this is used as image or registry, Quarkus throws exception.

Please add support for full image path in container-image configuration.

Exception:

```
 [ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.4.0.CR1:build (default) on project asc-api-service: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
 [ERROR] 	[error]: Build step io.quarkus.container.image.jib.deployment.JibProcessor#buildFromJar threw an exception: java.lang.IllegalArgumentException: The supplied container-image registry 'gitlab.company.hr:1111/group/subgroup/project' is invalid
 [ERROR] 	at io.quarkus.container.image.jib.deployment.JibProcessor.getImageReference(JibProcessor.java:190)
 [ERROR] 	at io.quarkus.container.image.jib.deployment.JibProcessor.createContainerizer(JibProcessor.java:138)
 [ERROR] 	at io.quarkus.container.image.jib.deployment.JibProcessor.containerize(JibProcessor.java:121)
 [ERROR] 	at io.quarkus.container.image.jib.deployment.JibProcessor.buildFromJar(JibProcessor.java:79)
 [ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
 [ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
 [ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
 [ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
 [ERROR] 	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:931)
 [ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
 [ERROR] 	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
 [ERROR] 	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
 [ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
 [ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
 [ERROR] 	at java.base/java.lang.Thread.run(Thread.java:834)
 [ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

**Implementation ideas**

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #8588 

Also, clean up some technical debt related to the container-image extensions.
The handling of the image string and its parts is now much more consistent
between extensions

### CODE REVIEW

**Comment by gsmet** [comment]
I won't comment on the code but I think we need the particular GitLab case to be fully documented.

**Comment by geoand** [comment]
> I won't comment on the code but I think we need the particular GitLab case to be fully documented.

Yeah that makes sense.

I'll do that after I fix the actual test failures :)

**Comment by geoand** [comment]
Tests fixed and doc added

File: extensions/container-image/spi/src/main/java/io/quarkus/container/spi/ImageReference.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.google.common.base.Strings;
+
+/**
+ * This is basically a simplified version of {@code com.google.cloud.tools.jib.api.ImageReference}
+ */
+public class ImageReference {
```
Comment (iocanel) [inline extensions/container-image/spi/src/main/java/io/quarkus/container/spi/ImageReference.java:27]: +1 we might want to replace our ImageUtils with that.

**Review by iocanel — APPROVED** [review]
LGTM

File: extensions/container-image/spi/src/main/java/io/quarkus/container/spi/ImageReference.java (hunk 1 above)
Comment (geoand) [reply extensions/container-image/spi/src/main/java/io/quarkus/container/spi/ImageReference.java:27]: Good idea! 

Let's do that in a follow up. Perhaps the PR you are working on that reworks S2I?

**Pull request merged**